Add go-for comprehension, go-try> for error channels.

whilo opened the pull request:
Also extend some docs, add more tests and some notes. I would like to have some standardized way to do error handling and higher level operations on channels and hence developed similar macros (1). I have started to reimplement the for-comprehension with channels instead of seqs to significantly reduce boilerplate code fiddeling around the function-boundary limits of core.async (similar to your reduce>) (2). The for-comprehension is still not ideal, as it is mostly eager atm.
My needs are a little bit different than yours, since I need these macros to work cross-platform between Clojure and cljs. This would be my next step, which is not totally trivial because of the way cljs handles macros (which is also changing atm. as it is becoming self-hosted). Are you aware of other libraries wrapping similar core.async functionality?

In general I would like to see some explicit syntactic pattern definition for the combination of <, >, !, ? and \* to compose macro names for channel operations consistently and make sure that all combinations are consistently implemented. It would also be nice to integrate an erlang-like actor library, so that the macros prove consistent in a larger context. This is just an idea though.

Cheers,
Christian

(1) https://github.com/whilo/replikativ/blob/master/src/replikativ/error.cljc
(2) https://github.com/whilo/replikativ/blob/master/src/replikativ/go_for.cljc

Commit:
Add go-for comprehension and supervision. Includes tests and docs.

Commit:
Add go-for comprehension and supervision. Includes tests and docs.

k7d commented:
Btw we have functional equivalent now for ClojureScript as well: https://github.com/fullcontact/full.cljs/tree/master/full.cljs.async

We first tried to combine it in one codebase but unfortunately it got too convoluted due to the way ClojureScript deals with macros.

k7d commented on full.async/src/full/async.clj:
```diff
@@ -62,6 +68,26 @@
              (recur (dec retries#)))
            res#)))))
 
+(defmacro go-try> [err-chan & body]
```
I'm wondering if this could be combined into go-try? The macro could check if first argument is a channel, and use that as error channel, otherwise work as before?

Btw seems that we haven't documented it anywhere but the way we use > suffix on functions is to indicate that function returns a channel with single result. For functions that return a channel with multiple results, we use >> suffix.

A kind of exception to this is go-try/go-retry but it's a macro and 'go' prefix already makes it clear that it returns a channel. What was your thinking with > suffix?

whilo replied:
You don't know whether the first argument expands to a channel at macro-expansion time, do you?

k7d replied:
Right. What's your thinking on having a separate error channel anyway?

whilo replied:
The general motivation is to have sane error management, as it can be very hairy in distributed systems. As Rich suggested, I had a look at error handling in erlang and as far as I understood I think it is well designed. The problem with go-loops et al. is that they are no named identities in your system like actors, so you cannot monitor them by address. I think the best way to decouple parts of the system then is by giving them a dedicated error channel to report on (with the help of the *-try macros) and then monitor for errors at subsystem boundaries and restart parts of the system as needed. That way one can start with a single dedicated error-channel, pass it in all systems to catch all errors (like a distributed try-catch) and then make the system robust by catching errors at appropriate boundaries and restart parts of the system.

For a single go routine it might not be as important (go-try>) as I usually block on it and its error anyway, I added it for consistency. But you cannot track errors in go-loops easily if they run concurrently.  These ideas are still wip, what is your take on error handling?

k7d replied:
What we did with go-try and <? was to introduce the minimum viable approach to error handling. Since anything in Java can throw the way core.async deals with exceptions isn't really sufficient for production-level code.

When we use go-try and <? everywhere, the exception normally gets popped to appropriate level in the system where it gets handled appropriately.

This is of course really just mimics Java's general approach to error handling.

k7d commented on full.async/src/full/async.clj:
```diff
@@ -78,6 +104,7 @@
         ; doall to check for throwables right away
         (doall)))
 
+;; TODO lazy-seq vs. vector in <<! ?
```
What do you mean? Lazy-seq is a big advantage here (which we leverage in our internal code for lazy loading various paged http resources). I guess it might be possible to make <<! return lazy-seq as well.

whilo replied:
Sure, this is an option. I was just looking for consistency issues between double-bang and single-bang versions.

k7d replied:
I remember I couldn't figure out an easy way to make <<! lazy but probably didn't spend enough time thinking about it.

whilo replied:
Sadly it is not possible atm. because lazy-seqs are lazy for the caller and need blocking io inside a single function application. core.async operations block inside a go-routine (block) and cannot block at function boundaries. That is also why I needed to change the for-comprehension over seqs to one over channels. The only way to overcome this would be by implementing some way to park functions at runtime level like the go-runtime or http://docs.paralleluniverse.co/pulsar/ does.

k7d commented on full.async/src/full/async.clj:
```diff
@@ -178,3 +205,77 @@
           (async/close! result))))
     result))
+
+;; taken from clojure/core ~ 1.7
```
this is probably redudant as master is now bumped to 1.7?

whilo replied:
This was just to show that the macro code is coming from core.clj originally. I have had to change it quite a bit anyway.

k7d commented:
@whilo what do you think of merging go-try> into go-try and go-loop-try> into go-loop-try per suggestion? other than that lgtm

whilo commented:
What about (go-try :error ch ~body) and (go-try-loop :error ch ~bindings ~body)? Maybe :error is not perfect. Could you document your syntactic scheme in the README btw.? I would like to extend this then to the error handling and als add some default monitoring routines.

k7d commented:
How about using bindings like this:
```clojure
(with-error-chan x
  (go-try
    ; ...
    ))
```

whilo commented:
Another semantic incosistency between <<!! is that <<! is eager and only returns once all values are put into the the vector btw.

whilo commented:
Oh, I just saw your bindings suggestion. I have to think about it.

k7d commented:
The binding-way may actually be more covenient as you wouldn’t need to pass around error channel at every level - just handle it at system / error handling boundaries.  


On Monday, August 17, 2015 at 4:40 PM, Christian Weilbach wrote:

> Oh, I just saw your bindings suggestion. I have to think about it.
>  
> —
> Reply to this email directly or view it on GitHub (https://github.com/fullcontact/full.monty/pull/9#issuecomment-131984268).
>  
>  
>  
>  
>

whilo commented:
http://stuartsierra.com/2013/03/29/perils-of-dynamic-scope 

You are right that binding might be sufficient. There are some problems I would like to address:

1. Never miss an exception. Since the mechanism is not built into the language (like try-catch), when you miss the binding somewhere you might see nothing. This is a general problem that can be there with go-try in general. 

2. Don't get weird multi-threading issues as all bindings are thread-local. While `core.async` copies the bindings between threads, one can easily be screwed by some multithreading somewhere.

3. A problem is that you need to know whether the exception will bubble with `(<? (go-try ...))` or be dealt with an error channel. If we do both in the same macros then each macro has different error handling semantics dependent on the runtime, e.g. try/catch will not catch local exceptions because the macro puts it on the error-channel of the binding. The `(go-try :error ...)` syntax is explicit at the call site.

4. The binding just makes it easier to get unintended interactions between different levels of bindings as it is not as explicit and allows error-handling to be an after thought (which is bad for robustness).

In general the binding is similar to how try/catch decouples error throwing from catching. 1. can be (partially) solved by having an annoying default binding (e.g. printing in the REPL/stderr). 2. is a problem we either have to leave to the user or cannot really address. 3. might be dealt with different macros for each case. I am not sure about 4. I think we need to formalize some error handling concept we expect people to use. I would try to copy the Erlang concept, if we have no better ideas.  Do you have experience with Go? 

Here is a somewhat related discussion http://dev.clojure.org/display/design/Resource+Scopes and maybe we should mention the problem at least on the clojure list so experienced people could chime in as it is a constant topic.

whilo commented:
See for instance http://blog.erlware.org/some-thoughts-on-go-and-erlang/ and https://blog.golang.org/errors-are-values
I think the Erlang (1) concept is both closer to Clojure/core.async semantics with throw/catch and I think the simple error returning in go is prone to treat errors wrong locally instead of allowing a whole process context to fail. 
(1) http://www.erlang.org/download/armstrong_thesis_2003.pdf

whilo commented:
As far as discussion and thinking have come so far, two hard requirements for robust error handling are necessary:

- *All* errors (exceptions) need to be caught and propagated. 
- Resources need to be freed.

In Erlang this is achieved by two mechanisms. Linking of processes by shared error-messages and monitoring allow to get error messages of the linked processes and to monitor (trap) those to restart parts of the system if some uncaught error occurs. This requires bidirectional communication, so a single error channel is not enough (unless one uses closing as a special signal, which has limitations). Furthermore Erlang can probably terminate processes preemptively, because this error signaling is built into the runtime (via mailboxes and their treatment of errors). We cannot do this without deep code transformations. But we can get probably a good compromise without adding manual cumbersome `abort` mechanisms, by just rethrowing a special `abort` exception on each blocking routine `<?`, `>?`, `alt?`, etc. That way each go-loop or go-process receives the abort signal through a locally thrown exception and can clean up through the usual try-catch-finally mechanism. This means all subsystem contexts propagate the abort exception everywhere to free resources and terminate all go-processes. For this to work reliably all code needs to use `>?`, `<?`, `alt?` and go-try routines which do not allow to swallow exceptions. Therefore I have adapted the go-try routines to either take a pair of `[exception-channel abort(-channel)]` or to rethrow by default, e.g. `<?go-try`.

What do you think? I have thought about the binding, but the thread-locality makes it too easy to mess the error-handling bindings up somewhere in my opinion.

k7d commented:
Hey @whilo my concern now is that go-try is not backwards compatible. How about your earlier suggestion of having optional keyword args ie. (go-try :error ch ~body)

This way (go-try ...) would behave as before, and error/abort channels would be optional.

whilo commented:
Well, optional can be dangerous. If you want to be robust your mechanism needs to catch all errors by default. If you put a `go-try` somewhere without `<?` then the exception is swallowed. Since one cannot force blocking on the return value I now track exceptions in `pending-errors` and at least print a warning on stale exceptions in the REPL (this is still not good). So `go-try` behaves again as before, but one can screw up. I currently see no way to fix this besides extending `core.async`. I also tried to use bindings, but as expected the thread-local binding depends on the timing of execution of the go-routines and is undeterministic (read: fundamentally broken).

 `delay-supervisor` should cover your ``go-retry functionality, but also terminates all nested go-loops freeing the resources and restarts the whole subsystem. See the final comment block. For this to work reliably all code needs to use `<?`, `>?` and `alt?`. It is still not able to detect when all operations are finished, so it uses a timeout at the moment. What do you think?

whilo commented:
I know have build a proper protocol and the supervisor waits until all subroutines are finished, so it should be pretty robust. I have also moved the tracking of stale exception into the supervision context. See the comment block at the end of the file for an example. You can play around with the timeouts, esepecially the one before the foo exception and reduce it to <100ms, then you will see foo propagate instead of the stale exception.

whilo commented:
I have completed the tests, added documentation and think the concept is reasonably worked out to be merged. Do you have any further suggestions?

k7d commented:
Allright looks solid. Let me try to resolve conflicts and merge this.

k7d commented:
Actually it might be easier if you do rebase on your end.

Theres are two new commits that are not in your branch:
https://github.com/fullcontact/full.monty/commit/dec898704a215000c6de5e7be4c9e9abfbb8c93f#diff-eda65a886d323fdf8bf36ff1a1b72f00
https://github.com/fullcontact/full.monty/commit/ef0b5c8ee384d1cb0f7e03f33dd8e30029fd93b8

whilo commented:
Like this?

whilo commented:
Seems like concat>> got lost.

whilo commented:
I will fix that later, my first rebase :) ...

Commit:
Cover explicit delay in supervisor like go-retry.

Commit:
Cover explicit delay in supervisor like go-retry.

whilo commented:
Ready to get merged! :D

Commit:
Almost make restarting-supervisor drop-in replacement for go-retry.

Commit:
Almost make restarting-supervisor drop-in replacement for go-retry.

whilo commented:
I have copied your retry conditions, the supervisor behaves the same now. Is this usable for you?

whilo commented:
Actually go-retry can be rewritten as syntactic sugar for the supervisor. If you are interested, I can do it.

whilo commented:
Hey, anything I can do to get this merged? I really don't want to fork `full.async`.

blendmaster commented:
After doing slightly more clojure myself, I think I feel qualified to say this is mergeable (after rebase). However, given all the back and forth in this PR, I'd prefer is if you move `go-for` and the supervision stuff to a separate "experimental"  or "labs" namespace, so we have more flexibility to make breaking changes (or remove if it's not useful) than if it's in full.async directly (where it's presumably stable).

Commit:
Move supervision and go-for into lab namespace.

Commit:
Merge.

Commit:
Merge.

Commit:
Do homework on bindings and transparently carry it on.

Move as much supervision stuff in the lab namespace as possible.

whilo commented:
@k7d  was right, it is possible to use a thread-binding if you hand it through everywhere. I have done that now and this simplifies call signatures considerably. You can have a look at the tests, if you like. It would be nice if you could try to replace go-retry with a restarting-supervisor in some of your cases and test it a little bit, I will try to fix any problems you have. Any feedback is appreciated, otherwise I hope this can be merged.

Commit:
Cover some resource management on abortion.

whilo commented:
What is your current status regarding a merge? I would like to use the improved error handling in [replikativ](https://github.com/replikativ/replikativ) without creating a fork first. 

I guess a problem is the port to cljs? We will port it and I will also maintain the error handling as it is a dependency for replikativ. But even now it is backwards compatible, so if you don't use the supervisor, the code will stay platform neutral with the previous cljs full.async code. In general, what would be your preferred way of cooperation.

Commit:
Merge branch 'master' of github.com:fullcontact/full.monty

Commit:
Add cljs support.

Supervisor binding is broken with core.async and cljs atm.

whilo commented:
I have implemented clojurescript support with a joint codebase (as this keeps them in sync much better). This completes the feature set I need atm., but I found out that cljs binding is somewhat broken here: https://groups.google.com/forum/#!topic/clojure/6cmnkHmHBNw 

This only affects supervisor usage from the experimental lab namespace.

Commit:
Fix issues macros for cljs.

The midje tests only work for clj atm., so the macros for cljs are verified only
by usage in the replikativ stack atm. and might still have issues.

whilo commented:
Besides the binding issue in cljs everything is working now for me (in the replikativ stack). I have added some further tests where they missed broken macros. If you need anything from me for a merge (rebase, docs, changes...), let me know.

Xorlev commented:
@k7d @skazhy is this mergable?

Commit:
Print stack traces in cljs.

skazhy commented:
We have cljs version for `full.async` in a [separate repo](https://github.com/fullcontact/full.cljs). I don't think we'll be migrating `full.monty` to be usable in both JVM and JS lands, so it would be best that these things can be added in `full.cljs` as well.

whilo commented:
We know. In fact @kordano has a fork on github for `full.cljs.async` with go-for and some macros. `full.monty` shouldn't be cross-platform of course, but I think `full.async` should be, because `core.async` is and there are valid use cases to use it cross-platform and not just two separate versions of it. Keeping two separate code bases in sync can be really painful (I think this is why you have bundled `full.monty`) and it is important to us that `full.async` behaves the same in both Clojure and ClojureScript. I have to admit that having cross-platform macros is not totally trivial because of the ClojureScript macroexpansion is in Clojure, but I think this is improving and will improve. If you have specific problems with it, we can fix it. If you really think it is better to have two codebases for `full.async` I can pull it apart, but I'd prefer not to.

One problem I am working out atm. is the lack of a working binding mechanism in ClojureScript. https://gist.github.com/whilo/a8ef2cd3f0e033d3973880a2001be32a

k7d commented:
@whilo i'm starting to think of branching out full.async back to it's own dedicated repo which could host both clj and cljs versions. The reason cljs end it up in it's own namespace (and project) was as you mentioned because of cljs macros. I guess it's still doable but messy. Even core.async uses different namespaces for clj & cljs. But I agree, it makes more sense for both version to live in the same codebase / repo.

whilo commented:
@k7d Ok, just out of curiousity, what is the problem with having it in `full.monty`? I am not familiar with joint leiningen repositories. 

I think that cljs macros are bit ugly and I think it could be better if we did not have separate `cljs.core.async` namespaces, which allowed to use the same macro code for both platforms. You also for instance have to include your own namespace in cljs for the clj macros. But it is still a bearable overhead for having an awesome cross platform async experience. It is also nice that the `full.async` macros will use the same namespaces with a joint codebase. Other projects like `datascript` use it as well, so one can have a look at how other solve problems (e.g. the if-cljs function).

k7d commented:
> Ok, just out of curiousity, what is the problem with having it in full.monty? I am not familiar with joint leiningen repositories.

Technically none, it's mostly just non-technical considerations:
* full.async doesn't have dependencies on anything else in full.monty
* full.async is probably the most interesting standalone part of full.monty
* nothing else in full.monty has / makes sense to have cljs support
* easier to manage contributions etc
* easier to discover, less confusion

k7d commented:
@whilo btw do you have a real world application of supervisor proposal? I get the theory but i'm very reluctant of pulling into master because it complicates the codebase. 

The part of full.async that is in master is all directly used in robust production code and directly driven by those requirements. It's a bit risky to pull in somewhat deep changes without having a clear understanding of it's practical value.

blendmaster commented:
I'll again suggest moving the supervisor stuff to a separate `full.async.labs`/`full.async.beta` namespace, so it can still be in releases without the same Satisfaction Guarantee™ the rest of the code has.

whilo commented:
@k7d I am using it everywhere in <https://github.com/replikativ/replikativ> and this is used e.g. in <https://topiq.es>. This is the whole motivation for it, since I need robust error handling in a p2p setting. I can understand that it is not paramount for you, because you always supervise all your servers. Still automatic recovery in case of error can be a relief. I also tried to add non-trivial unit tests for it, have you seen them?

@blendmaster I have moved as much as possible in the `lab` namespace as you suggested. I could move the protocol definitions themselves out in another namespace as well, but `go-try` and `<?` need to talk to the supervisor, so no exception slips out. Note that the `alts!` in `<?` does nothing by default since the default supervisor only prints stale errors, but never aborts anything. `go-try` only captures and restores the supervisor binding and tracks exceptions through it (in an atom). The behaviour should be completely unchanged. I could create another macro `go-try2` or something, but this would defeat the purpose of having everything supervised automatically. 

On what level would you be interested in supervision? I tried for instance to make the supervisor a drop-in replacement for your `go-retry`. 

I have ported topiq to the new supervisor, but I got a problem with broken bindings in cljs :/ http://dev.clojure.org/jira/browse/CLJS-1634 I will still be able to work around this, but this also effects the cljs port of full.async. If the issue was rejected, I needed to use lexical scoping (an explicit argument to go-try) to pass through the supervisor everywhere.

whilo commented:
Do you plan any major changes to `full.async`? I would propose to wait whether the Jira ticket is accepted/rejected and then merge the pull request, either with the current dynamical binding or with the original lexical binding for the supervisor. I will try to keep the fork up-to-date with your changes. 
But it will still be helpful to know whether and how you are interested in supervision. I guess it is not too important, but I would like to have a sound concept which can be used by default in full.async, so libraries of `full.async` users compose.

Commit:
Default to infinite retry, fix test.

Commit:
Small changes.

Commit:
Remove bug reproducing code.

Commit:
Push new minor version.

Commit:
Configurable logging for the supervisor.

Commit:
Use lexical binding everywhere.

Dynamic binding does not work in cljs and has a performance overhead in
clj and cljs.

skazhy commented:
Closing this in favor of [superv.async](https://github.com/replikativ/superv.async). Congrats on the release!

whilo commented:
Thanks. Maybe it will be mergeable again with full.async at some point. Thanks for all the work put into full.async!
